File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis.VisualBasic.Emit/TypeMemberReference.cs
using System.Collections.Generic;
using Microsoft.Cci;
using Microsoft.CodeAnalysis.Emit;
using Microsoft.CodeAnalysis.Symbols;
using Roslyn.Utilities;

namespace Microsoft.CodeAnalysis.VisualBasic.Emit
{
	internal abstract class TypeMemberReference : ITypeMemberReference
	{
		protected abstract Symbol UnderlyingSymbol { get; }

		private string INamedEntityName => UnderlyingSymbol.MetadataName;

		public virtual ITypeReference GetContainingType(EmitContext context)
		{
			return ((PEModuleBuilder)context.Module).Translate(UnderlyingSymbol.ContainingType, (VisualBasicSyntaxNode)context.SyntaxNode, context.Diagnostics);
		}

		ITypeReference ITypeMemberReference.GetContainingType(EmitContext context)
		{
			//ILSpy generated this explicit interface implementation from .override directive in GetContainingType
			return this.GetContainingType(context);
		}

		public override string ToString()
		{
			return UnderlyingSymbol.ToDisplayString(SymbolDisplayFormat.ILVisualizationFormat);
		}

		private IEnumerable<ICustomAttribute> IReferenceAttributes(EmitContext context)
		{
			return SpecializedCollections.EmptyEnumerable<ICustomAttribute>();
		}

		IEnumerable<ICustomAttribute> IReference.GetAttributes(EmitContext context)
		{
			//ILSpy generated this explicit interface implementation from .override directive in IReferenceAttributes
			return this.IReferenceAttributes(context);
		}

		public abstract void Dispatch(MetadataVisitor visitor);

		void IReference.Dispatch(MetadataVisitor visitor)
		{
			//ILSpy generated this explicit interface implementation from .override directive in Dispatch
			this.Dispatch(visitor);
		}

		private IDefinition IReferenceAsDefinition(EmitContext context)
		{
			return null;
		}

		IDefinition IReference.AsDefinition(EmitContext context)
		{
			//ILSpy generated this explicit interface implementation from .override directive in IReferenceAsDefinition
			return this.IReferenceAsDefinition(context);
		}

		private ISymbolInternal IReferenceGetInternalSymbol()
		{
			return UnderlyingSymbol;
		}

		ISymbolInternal IReference.GetInternalSymbol()
		{
			//ILSpy generated this explicit interface implementation from .override directive in IReferenceGetInternalSymbol
			return this.IReferenceGetInternalSymbol();
		}

		public sealed override bool Equals(object obj)
		{
			throw ExceptionUtilities.Unreachable;
		}

		public sealed override int GetHashCode()
		{
			throw ExceptionUtilities.Unreachable;
		}
	}
}

File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis/VisualBasic/Emit/TypeMemberReference.vb
Imports Microsoft.Cci
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.Emit
Imports Microsoft.CodeAnalysis.Symbols
Imports Microsoft.CodeAnalysis.VisualBasic
Imports Roslyn.Utilities
Imports System
Imports System.Collections.Generic

Namespace Microsoft.CodeAnalysis.VisualBasic.Emit
	Friend MustInherit Class TypeMemberReference
		Implements ITypeMemberReference
		ReadOnly Property INamedEntityName As String
			Get
				Return Me.UnderlyingSymbol.MetadataName
			End Get
		End Property

		Protected MustOverride ReadOnly Property UnderlyingSymbol As Symbol

		Protected Sub New()
			MyBase.New()
		End Sub

		Public MustOverride Sub Dispatch(ByVal visitor As MetadataVisitor)

		Public NotOverridable Overrides Function Equals(ByVal obj As Object) As Boolean
			Throw ExceptionUtilities.Unreachable
		End Function

		Public Overridable Function GetContainingType(ByVal context As EmitContext) As ITypeReference Implements ITypeMemberReference.GetContainingType
			Return DirectCast(context.[Module], PEModuleBuilder).Translate(Me.UnderlyingSymbol.ContainingType, DirectCast(context.SyntaxNode, VisualBasicSyntaxNode), context.Diagnostics, False, False)
		End Function

		Public NotOverridable Overrides Function GetHashCode() As Integer
			Throw ExceptionUtilities.Unreachable
		End Function

		Private Function IReferenceAsDefinition(ByVal context As EmitContext) As IDefinition
			Return Nothing
		End Function

		Private Function IReferenceAttributes(ByVal context As EmitContext) As IEnumerable(Of ICustomAttribute)
			Return SpecializedCollections.EmptyEnumerable(Of ICustomAttribute)()
		End Function

		Private Function IReferenceGetInternalSymbol() As ISymbolInternal
			Return Me.UnderlyingSymbol
		End Function

		Public Overrides Function ToString() As String
			Return Me.UnderlyingSymbol.ToDisplayString(SymbolDisplayFormat.ILVisualizationFormat)
		End Function
	End Class
End Namespace